src/lib/stream/shake_cipher/shake_cipher.cpp

```diff
@@ -62,6 +62,44 @@ void SHAKE_Cipher::cipher(const uint8_t in[], uint8_t out[], size_t length)
    m_buf_pos += length;
    }
 
+void SHAKE_Cipher::write_keystream(uint8_t out[], size_t length)
+   {
+   verify_key_set(m_state.empty() == false);
+
+   if(m_buf_pos > 0)
+      {
+      const size_t take = std::min(length, m_shake_rate - m_buf_pos);
+      copy_mem(out, &m_buffer[m_buf_pos], take);
+      out += take;
+      length -= take;
+      m_buf_pos += take;
+
+      if(m_buf_pos == m_shake_rate)
+         {
+         SHA_3::permute(m_state.data());
+         m_buf_pos = 0;
+         }
+      }
+
+   if(length == 0)
+      return;
+
+   BOTAN_ASSERT_NOMSG(m_buf_pos == 0);
+
+   while(length >= m_shake_rate)
+      {
+      copy_out_le(out, m_shake_rate, m_state.data());
+      SHA_3::permute(m_state.data());
+      length -= m_shake_rate;
+      out += m_shake_rate;
+      }
+
+   copy_out_le(m_buffer.data(), m_shake_rate, m_state.data());
+
+   copy_mem(out, &m_buffer[0], length);
+   m_buf_pos += length;
+   }
+
```

Comment:
Potentially, a similar refactoring as for CTR could be used, didn't look into that so far.

Reply:
See other comment